templates/company-board/contracts/CompanyBoardTemplate.sol

```diff
@@ -48,42 +38,92 @@ contract CompanyBoardTemplate is BaseTemplate {
         _ensureMiniMeFactoryIsValid(_miniMeFactory);
     }
 
-    function prepareInstance() public {
+    function prepareInstance(string _shareTokenName, string _shareTokenSymbol) public {
         (Kernel dao,) = _createDAO();
         MiniMeToken boardToken = _createToken(BOARD_TOKEN_NAME, BOARD_TOKEN_SYMBOL, BOARD_TOKEN_DECIMALS);
-        MiniMeToken shareToken = _createToken(SHARE_TOKEN_NAME, SHARE_TOKEN_SYMBOL, SHARE_TOKEN_DECIMALS);
+        MiniMeToken shareToken = _createToken(_shareTokenName, _shareTokenSymbol, SHARE_TOKEN_DECIMALS);
         _storeCache(dao, boardToken, shareToken, msg.sender);
     }
 
-    function setupInstance(string _id, address[] _boardMembers, address[] _shareHolders, uint256[] _shareStakes) public {
+    function setupInstance(
+        string _id, 
+        address[] _boardMembers, 
+        address[] _shareHolders, 
+        uint256[] _shareStakes,
+        uint64[] _boardVoteSettings, /* [voteDuration, supportRequired, minAcceptanceQuorum] */
+        uint64[] _shareVoteSettings, /* idem */
+        uint64 _financePeriod
+    ) 
+        public 
+    {
+        require(_boardVoteSettings.length == 3, ERROR_BAD_VOTE_SETTINGS);
+        require(_shareVoteSettings.length == 3, ERROR_BAD_VOTE_SETTINGS);
+        (Agent agent, Finance finance) = _setupCommon(_financePeriod);
+        (TokenManager boardTokenManager, Voting boardVoting) = _setupBoard(_boardMembers, _boardVoteSettings[0], _boardVoteSettings[1], _boardVoteSettings[2]);
+        (TokenManager shareTokenManager, Voting shareVoting) = _setupShare(_shareHolders, _shareStakes, _shareVoteSettings[0], _shareVoteSettings[1], _shareVoteSettings[2]);
+        _setupPermissions(boardVoting, boardTokenManager, shareVoting, shareTokenManager, agent, finance);
+        _registerDAO(_id);
```

Comment:
I'm sure we can optimize the caching here a little bit. Probably you can pop up the cache at the beginning and pass the dao/acl by parameter to the rest of the functions instead of loading the cache on each sub function

Reply:
Yeah, I had tried this before but it "caused stack too deep" errors. I uploaded that version in case you want to see. I failed at the linked line: https://github.com/aragon/dao-templates/blob/parameterize_templates_simpler_cache/templates/company-board/contracts/CompanyBoardTemplate.sol#L65

Reply:
Ofc that indexing variables at that level won't help, you should pass the array to the internal function and check it there

Reply:
Yes, you're right! Still, I just tried changing that, and the stack too deep error persists. I've pushed it to that temporary branch nonetheless, so that we may use this code as reference for a future gas optimization.